qutip/tests/test_qip.py

```diff
@@ -0,0 +1,45 @@
+import sys
+
+import pytest
+
+
+class QutipQipStub:
+    class CircuitModuleStub:
+        QubitCircuit = "FakeQubitCircuit"
+
+    circuit = CircuitModuleStub()
+
+
+@pytest.fixture
+def without_qutip_qip(monkeypatch):
+    monkeypatch.setitem(sys.modules, "qutip_qip", None)
+    monkeypatch.delitem(sys.modules, "qutip.qip", raising=False)
+
+
+@pytest.fixture
+def with_qutip_qip_stub(monkeypatch):
+    monkeypatch.setitem(sys.modules, "qutip_qip", QutipQipStub())
+    monkeypatch.delitem(sys.modules, "qutip.qip", raising=False)
+
+
+def test_failed_import(without_qutip_qip):
+    # Ensure 'qutip.qip' is not imported yet
+    assert "qutip.qip" not in sys.modules
+    with pytest.raises(
+        ImportError,
+        match="Importing 'qutip.qip' requires the 'qutip_qip' package.",
+    ):
+        import qutip.qip
+
+
+def test_with_qip(monkeypatch, with_qutip_qip_stub):
+    import qutip.qip
+
+    monkeypatch.setitem(sys.modules, "qutip.qip.circuit", qutip.qip.circuit)
+    import qutip.qip.circuit as circuit
+    from qutip.qip.circuit import QubitCircuit
+    import qutip_qip
```

Comment:
I agree that it would be nice to test that sub-modules are also importable (i.e. does our approach handle that case) but I don't think monkeypatching sys.modules in the test shows that has has worked.

I guess a much better test is to instead of creating a dummy `qutip_qip` here and sticking it in sys.modules, is to, in the `with_qutip_qip_stub` fixture to use `tempfile` to create a temporary folder, add that temporary folder temporarily to sys.path, and write into the temporary folder a very tiny package with modules in it.

Reply:
That's an ingenious idea! I gave it a shot, let me know what you think